enarx-pr-assign

```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/python3
+# SPDX-License-Identifier: Apache-2.0
+
+from github import Github
+import os
+
+RESPONSIBILITIES = {
+    'CHANGES_REQUESTED': lambda author, reviewer, requests: reviewer if reviewer in requests else author,
+    'DISMISSED': lambda author, reviewer, requests: reviewer,
+    'APPROVED': lambda author, reviewer, requests: None,
+    'COMMENTED': lambda author, reviewer, requests: reviewer,
+    'PENDING': lambda author, reviewer, requests: reviewer,
+}
+
+github = Github(os.environ['GITHUB_TOKEN'])
+
+def handle(issue, pr, responsible):
+    assignees = {a.login for a in pr.assignees}
+
+    remove = assignees - responsible
+    needed = responsible - assignees
+
+    repo = issue.repository
+    print(f"{repo.organization.login}/{repo.name}#{pr.number}:", end='')
+    for u in needed:
+        print(f" +{u}", end='')
+    for u in remove:
+        print(f" -{u}", end='')
+    print()
+
+    if remove:
+        pr.remove_from_assignees(*remove)
+    if needed:
+        pr.add_to_assignees(*needed)
+
+for issue in github.search_issues(f"org:enarx is:pr is:public is:open -is:draft"):
+    pr = issue.as_pull_request()
+
+    # Get the latest reviews for all PR reviewers.
+    latest_reviews = {r.user.id: r for r in pr.get_reviews()}
+    # Pick out the latest Changes Requested reviews from above.
+    changes_requested_reviews = {u.user.id: u for (k, u) in latest_reviews.items() if u.state == "CHANGES_REQUESTED"}
+
+    # Collect reviewers whose Changes Requested reviews have been addressed in a set.
+    addressed_reviewers = {e.requested_reviewer.login for e in pr.get_issue_events()
+        if e.requested_reviewer is not None and
+        e.requested_reviewer.id in changes_requested_reviews and
+        e.created_at > changes_requested_reviews[e.requested_reviewer.id].submitted_at}
+
+    # Get requested reviewers that have not yet submitted any form of review.
+    pending_requested_reviewers = {r for r in pr.get_review_requests()[0] if r.id not in latest_reviews}
+    
+    # Combine the two prior sets to form the set of all reviewers that have been
+    # requested for review on the PR.
+    requests = pending_requested_reviewers | addressed_reviewers
+    
+    # Construct the set of responsible people for this PR.
+    responsible = requests | {RESPONSIBILITIES[r.state](pr.user.login, r.reviewer.login, requests) for (k, r) in latest_reviews}
+    responsible.remove(None)
+    
+    # Handle the set of responsible people.
+    handle(issue, pr, responsible)
```

Comment:
Let's break this loop up a bit. We probably want `get_latest_reviews(pr)` and `get_requested_reviewers(pr)` functions. That will cut down on the intermediate variables we have to think about. Also, the `get_requested_reviewers(pr)` function can probably be implemented with `yield`, making the logic easier to think about.

Reply:
Done in latest revision.